nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java

```diff
@@ -0,0 +1,141 @@
+import java.util.Map;
+import org.apache.iceberg.BaseMetastoreTableOperations;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.io.FileIO;
+
+/**
+ * Nessie implementation of Iceberg TableOperations.
+ */
+public class NessieTableOperations extends BaseMetastoreTableOperations {
+
+  private final NessieClient client;
+  private final ContentsKey key;
+  private UpdateableReference reference;
+  private IcebergTable table;
+  private FileIO fileIO;
+
+  /**
+   * Create a nessie table operations given a table identifier.
+   */
+  public NessieTableOperations(
+      ContentsKey key,
+      UpdateableReference reference,
+      NessieClient client,
+      FileIO fileIO) {
+    this.key = key;
+    this.reference = reference;
+    this.client = client;
+    this.fileIO = fileIO;
+  }
+
+  @Override
+  protected void doRefresh() {
+    try {
+      reference.refresh();
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException("Failed to refresh as ref is no longer valid.", e);
+    }
+    String metadataLocation = null;
+    try {
+      Contents contents = client.getContentsApi().getContents(key, reference.getHash());
+      this.table = contents.unwrap(IcebergTable.class)
+          .orElseThrow(() ->
+              new IllegalStateException("Cannot refresh iceberg table: " +
+                  String.format("Nessie points to a non-Iceberg object for path: %s.", key)));
+      metadataLocation = table.getMetadataLocation();
+    } catch (NessieNotFoundException ex) {
+      if (currentMetadataLocation() != null) {
+        throw new NoSuchTableException(ex, "No such table %s", key);
+      }
+    }
+    refreshFromMetadataLocation(metadataLocation, 2);
+  }
+
+  @Override
+  protected void doCommit(TableMetadata base, TableMetadata metadata) {
+    reference.checkMutable();
+
+    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+
+    boolean threw = true;
+    try {
+      IcebergTable newTable = ImmutableIcebergTable.builder().metadataLocation(newMetadataLocation).build();
+      client.getContentsApi().setContents(key,
+                                          reference.getAsBranch().getName(),
+                                          reference.getHash(),
+                                          String.format("iceberg commit%s", applicationId()),
+                                          newTable);
+      threw = false;
+    } catch (NessieConflictException ex) {
+      String fixMsg = reference.isBranch() ?
+          String.format("Update the reference %s and try again", reference.getName()) :
+          String.format("Can't commit to the tag %s", reference.getName());
+      throw new CommitFailedException(ex, "Commit failed: Reference hash is out of date. %s", fixMsg);
```

Comment:
`CommitFailedException` is used to trigger a table refresh and a retry. Throwing it for `NessieConflictException` seems correct to me, but reading the error messages makes me less sure.

If the ref is a tag, then we don't want to retry because it can't be updated, right? In that case, this should throw some other exception because the table is read-only.

If the ref is a branch, then the ref should be refreshed using the normal retry logic so everything looks good. `doRefresh` will update the ref and then update the table content.

Reply:
Yep! exactly! On line 87 we do `reference.checkMutable()` which will throw `IllegalArgumentException` if its a tag. So by the time we commit we know its a branch. I have cleaned up the erroneous error msg.